cfgrammar/src/lib/yacc/parser.rs

```diff
@@ -55,20 +55,76 @@ pub enum YaccParserErrorKind {
     DuplicateEPP(Vec<Span>),
     ReachedEOL,
     InvalidString,
+    ValidationNoStartRule,
+    ValidationInvalidStartRule,
+    ValidationUnknownRuleRef,
+    ValidationUnknownToken,
+    ValidationNoPrecForToken,
+    ValidationUnknownEPP,
 }
 
 /// Any error from the Yacc parser returns an instance of this struct.
-#[derive(Debug, PartialEq, Eq)]
+#[derive(Debug)]
 pub struct YaccParserError {
     pub kind: YaccParserErrorKind,
     pub span: Span,
+    pub sym: Option<Symbol>,
```

Comment:
So, my thinking is the right thing to do here, might be to just move this `sym` into the `Kind`, so these become e.g. `InvalidStartRule(Symbol)`, all these except `NoStartRule`, (which ignores sym) panic if `sym` is `None`,
so we could drop the `Option` then.

Reply:
So, that would basically fix *1.* and *3.*, one option for fixing *2.* is also getting rid putting the span field on `YaccParserError` moving it too into the variants, basically we'd have 3 cases:

```
IllegalInteger(Span)`
// Lots more Foo(Span)
DuplicateRule(Span, Vec<Span>)`
// Few more with (Span, Vec<Span>)
ValidationInvalidStartRule(Symbol),
// few more with `Symbol`
```

Which leaves `YaccParserError { kind: YaccParserErrorKind }`, at which point we can remove it `YaccParserError`,
and likely rename `YaccParserErrorKind` to an `YaccParserError`.

There is then some trouble because we no longer have a way to get at the span for an arbitrary error, but we wouldn't be duplicating it any longer, which the testsuite relies on.

I assume we'd just add an `fn span() -> Span` either as a trait but probably just as an `impl YaccParserError`.
I'd like to play around with this path, and see what the memory layout looks like.

Probably push something to look at in another branch starting from here, unless it sounds revolting.

Reply:
FWIW, here are a couple of commits implementing the above 2 comments.

https://github.com/ratmice/grmtools/commit/9ce3a49ea66a02d9ef6789c0ec4fc64514b867dd
https://github.com/ratmice/grmtools/commit/4c02011acb12a5fa1e12ad681472c866eddd8503

Reply:
It doesn't really seem terribly great memory layout, at `56`, if we limit it to `(Span, String)` the enum reduces to 48, the size of `Symbol`, It seems plausible that we can reduce it by making `ValidationInvalidStartRule(Rule)`, instead of using `Symbol`.

it may be that e.g. we return `Symbol` because some errors return `Rule` and some `Token`, while we could then use the specific type for each error, removing the need for the tag bits.

Reply:
Does the `(Span, String)` thing via a new `TokenSymbol` and `RuleSymbol`.
Anyhow I think that is as compact as it gets, and at `48` compared to `64` with `YaccGrammarError` + `GrammarValidationError` 

https://github.com/ratmice/grmtools/commit/9b6e932a0a9c8d240775a6fcd48f85676cb8ecfe